c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+    // within the fire() callback. The promise adapter can do this, because it lives outside of the
+    // coroutine's frame, unlike Coroutine<T> or Awaiter<U>.
+    node->onReady(this);
+  }
+
+  PromiseFulfiller<T>& fulfiller;
+  // Used to fulfill/reject the enclosing coroutine's kj::Promise<T>.
+
+private:
+  Maybe<Own<Event>> fire() override {
+    // The promise this coroutine is currently waiting on is ready. Extract its result and check for
+    // exceptions.
+
+    // Destroying or resuming the coroutine will leave these pointers dangling, so we call this
+    // function before either action.
+    auto nullifyPointers = [&]() {
+      node = nullptr;
+      result = nullptr;
+    };
+
+    node->get(*result);
+
+    // Note: the Promise::wait() implementation throws a recoverable exception if both a value and
+    // an exception are set. We only have visibility on the exception here, so the following might
+    // not have quite the same semantics.
+
+    // Check for exceptions. We could let Awaiter::await_resume() do it, but then we'd resume the
+    // coroutine just for it to rethrow, which seems pointless.
+    KJ_IF_MAYBE(exception, result->exception) {
+      nullifyPointers();
+      fulfiller.reject(kj::mv(*exception));
+      // The coroutine threw an exception. Our only choice is to destroy it.
+      //
+      // TODO(now): Should we ignore exceptions from destroy()?
+      coroutine.destroy();
+    } else {
+      nullifyPointers();
+      // Call Awaiter::await_resume() and proceed with the coroutine. Note that if this was the last
+      // co_await in the coroutine, control will flow off the end of the coroutine and it will be
+      // destroyed -- as part of the execution of coroutine.resume().
+      coroutine.resume();
+    }
+
+    return nullptr;
+  }
+
+  Coroutine::Handle coroutine;
+
+  PromiseNode* node = nullptr;
+  ExceptionOrValue* result = nullptr;
+  // Set by awaitSuspend(), used (and reset) in fire().
+};
+
+template <typename T>
+template <typename U>
+class Coroutine<T>::Awaiter {
+  // Wrapper around a co_await'ed promise and some storage space for the result of that promise.
+  // The compiler arranges to call our await_suspend() to suspend, and we forward the call to the
+  // enclosing coroutine's promise adapter, which arranges to be woken up when the awaited promise
+  // is settled. Once that happens, the enclosing coroutine's promise adapter resumes the coroutine,
+  // which transitively calls await_resume() to unwrap the awaited promise result.
+
+public:
+  explicit Awaiter(Promise<U> promise): promise(kj::mv(promise)) {}
+
+  bool await_ready() const { return false; }
+  // This could return "`promise.node.get()` is safe to call" instead, which would make suspension-
+  // less co_awaits possible for immediately-fulfilled promises. However, that's probably unwise,
+  // since it is explicitly not how .then() works.
```

Comment:
Hmm, I think this situation might be genuinely different from `.then()`. With `.then()`, you have this situation:

    promise = thing.then([]() { foo(); });
    bar();

And the question is: could `foo()` sometimes run before `bar()`. We want the answer to be consistent, and the only possible consistent answer is "no".

But with coroutines there is no possibility of a `bar()`. Once you invoke `await`, no further code is executed. So we don't have this scenario.

So the only question here is whether an `await` should always yield to the event loop. But KJ already goes out of its way to run most events in a LIFO-ish way, such that an immediate `.then()` actually _won't_ allow anything else to run in between, because we don't want to context switch too much. So in that case, it seems like enabling immediate `await`s is just a performance optimization and we might as well do it.

Reply:
Makes sense to me! And it would certainly be a nice optimization.

Reply:
Optimized: https://github.com/capnproto/capnproto/pull/1104/commits/b3c38109b0037493798ec11c24b5b6718669da64

However, I think that has a bug: `evalLater()` promises turn out to be immediately ready, so we breeze right through them. Since we frequently use `evalLater()` to mean "I really do want to yield to the event loop here," I figure we should come up with a solution to that.

Reply:
Hmm, I think you could solve this by changing `isArmed()` to `isNext()`, which would return true if the event is armed _and_ is next in line to be fired. In that case there's certainly no semantic difference.